### DISCUSSION TITLE
How to set different batching sizes for @stream

### DISCUSSION THREAD

**Discussion by gopeter** [urql-graphql/urql#2221] [Q&A] [answered]
I am building an app with a Rails GraphQL API and a React + urql frontend. Since I have to render very large lists (>3000 rows with > 30 columns), I wanted to try out the @stream directive. [There is a discussion about adding @stream to `graphql-ruby`](https://github.com/rmosolgo/graphql-ruby/discussions/3847) and an [example repo from the author of `graphql-ruby`](https://github.com/rmosolgo/graphql-ruby-urql-stream-example) which give a lot more information about how this is implemented on the Ruby side. Maybe this is useful for you too (taken from the Readme of the example repo):

> This took me forever to figure out:
> 
> Ruby sends a leading boundary (---) as part of the first patch. This is in the spec: there should be a boundary between the preamble part (with nothing in it, afaict) and the first patch.
> 
> At first, running the demo with Ruby server always looked "one patch behind." That is, the server would send the first patch, but nothing would render. Then it would send the second patch, and the first patch would render, then send the third patch, and the second patch would render. Weird!
> 
> I was fussing and fussing over trying to make the Ruby server exactly like the Javascript one. Finally, I noticed that the JS server has a res.write("---") before any patches, so I added that to the Ruby server (and removed that boundary from the first patch) and now it seems to work right.
> 
> So it seems like urql assumes that each flush of data from the server will contain exactly one patch.

But since everything is running fine on the server-side now, I've encountered another problem. Playing around with the timeouts of your `@stream` example show that `urql` does not render the list again for each item that arrives, but batches the renders between. Adding a timeout of 500ms between each alphabet char renders each char at once, but removing the timeout renders the whole list at once. Which is really great!

But looking at my list, the render happens at first after 3000 items are fetched:

![image](https://user-images.githubusercontent.com/1242960/151395163-3d03870c-7bf5-490a-8161-5c5ad32509d3.png)

I'm not sure why. Does `urql` has a threshold like "if we don't get a fresh update within 20ms, render the items we have"? Or how do you batch this? I would like to display the first 100 results and then stream the other ones in the background. But at the moment, the first render happens after almost all records are streamed.

**Comment by kitten** [comment, marked as answer]
We're not using any batching, actually; We trust that caching is fast enough to update itself immediately when a next chunk comes in and is sent along the exchange pipeline. So, actually, what I'd expect is that the batching doesn't come in from our end, but that something's going on in the render layer.

Have you tried to inspect the individual responses manually, or via the urql devtools (or via simple logging with the debug exchange)? It could also be that you're maybe using React's experimental concurrent mode?

I hope this isn't too convoluted, but you can actually see here that we're not doing any special debounces or delays: https://github.com/FormidableLabs/urql/blob/main/packages/core/src/internal/fetchSource.ts

**Reply by gopeter** [reply]
Oh okay 🤔 The only thing I did was to add `console.log(data.alphabet.length)` to https://github.com/FormidableLabs/urql/blob/main/examples/with-defer-stream-directives/src/Songs.jsx#L39 and to remove this line https://github.com/FormidableLabs/urql/blob/main/examples/with-defer-stream-directives/server/index.js#L32. Then I get just one render instead of rendering each char. That's why I thought that `urql` does some batching here 🙂
